asciidoc-confluence-publisher-client/src/main/java/org/sahli/asciidoc/confluence/publisher/client/ConfluencePublisher.java

```diff
@@ -187,43 +190,41 @@ private void updatePage(String contentId, String ancestorId, ConfluencePageMetad
         }
     }
 
-    private static String contentHash(String content) {
-        return sha256Hex(content);
-    }
-
     private void addAttachments(String contentId, Map<String, String> attachments) {
         attachments.forEach((attachmentFileName, attachmentPath) -> addOrUpdateAttachment(contentId, attachmentPath, attachmentFileName));
     }
 
     private void addOrUpdateAttachment(String contentId, String attachmentPath, String attachmentFileName) {
         Path absoluteAttachmentPath = absoluteAttachmentPath(attachmentPath);
+        String newContentHash = sha256Hash(fileInputStream(absoluteAttachmentPath));
 
         try {
             ConfluenceAttachment existingAttachment = this.confluenceClient.getAttachmentByFileName(contentId, attachmentFileName);
-            InputStream existingAttachmentContent = this.confluenceClient.getAttachmentContent(existingAttachment.getRelativeDownloadLink());
+            String attachmentId = existingAttachment.getId();
+            String existingContentHash = this.confluenceClient.getPropertyByKey(contentId, attachmentId);
 
-            if (!isSameContent(existingAttachmentContent, fileInputStream(absoluteAttachmentPath))) {
-                this.confluenceClient.updateAttachmentContent(contentId, existingAttachment.getId(), fileInputStream(absoluteAttachmentPath));
+            if (notSameHash(existingContentHash, newContentHash)) {
+                this.confluenceClient.deletePropertyByKey(contentId, attachmentId);
+                this.confluenceClient.updateAttachmentContent(contentId, attachmentId, fileInputStream(absoluteAttachmentPath));
+                this.confluenceClient.setPropertyByKey(contentId, attachmentId, newContentHash);
             }
 
         } catch (NotFoundException e) {
-            this.confluenceClient.addAttachment(contentId, attachmentFileName, fileInputStream(absoluteAttachmentPath));
+            ConfluenceAttachment newAttachment = this.confluenceClient.addAttachment(contentId, attachmentFileName, fileInputStream(absoluteAttachmentPath));
+            this.confluenceClient.setPropertyByKey(contentId, newAttachment.getId(), newContentHash);
         }
     }
 
     private Path absoluteAttachmentPath(String attachmentPath) {
         return Paths.get(attachmentPath);
     }
 
-    private static boolean notSameContentHash(String actualContentHash, String newContentHash) {
-        return actualContentHash == null || !actualContentHash.equals(newContentHash);
+    private static String contentHash(String content) {
+        return sha256Hex(content);
     }
 
-    private static boolean isSameContent(InputStream left, InputStream right) {
-        String leftHash = sha256Hash(left);
-        String rightHash = sha256Hash(right);
-
-        return leftHash.equals(rightHash);
+    private static boolean notSameHash(String actualContentHash, String newContentHash) {
```

Comment:
Related to above: we should agree on what "content" means: so far, we referred with it to the content of the page, so a page had a content and optionally attachments. So maybe here, using `actualHash` and `newHash` would be more appropriate (or any other name with more meaning, but including both content and attachments).